In [1]:
from google.cloud import bigquery
from google.oauth2 import service_account
import time
import matplotlib.pyplot as plt
import seaborn as sns
from google.cloud import bigquery_storage
import pandas as pd
import numpy as np
import plotly.graph_objects as go
import plotly.express as px
from itertools import combinations, islice
from collections import defaultdict
from tqdm import tqdm

# Explanation to Apriori Algorithm and Association Rules Mning

## Apriori Algorithm

The Apriori algorithm is a popular algorithm used for association rule mining in data mining and machine learning. It aims to discover relationships or associations between items in a dataset. The algorithm follows a "bottom-up" approach and uses a breadth-first search strategy to find frequent itemsets.

Here's a high-level overview of how the Apriori algorithm works:

Support: The algorithm begins by calculating the support of each item in the dataset. Support represents the frequency of occurrence of an item in the dataset.

Frequent Itemsets: Next, the algorithm generates a set of frequent itemsets based on a minimum support threshold set by the user. Frequent itemsets are subsets of items that meet the minimum support threshold.

Candidate Generation: The algorithm uses the frequent itemsets from the previous step to generate a new set of candidate itemsets. These candidate itemsets are formed by joining the frequent itemsets with themselves.

Pruning: The algorithm prunes the candidate itemsets that contain subsets that are not frequent. This step helps reduce the search space and improve the efficiency of the algorithm.

Repeat: Steps 3 and 4 are repeated iteratively until no new frequent itemsets can be generated.

## Association Rules Mining

Association rules mining is the process of discovering interesting relationships or associations among items in a dataset. It is based on the concept of conditional probability and is often used in market basket analysis, where the goal is to find associations between products frequently purchased together.

Association rules are typically represented in the form of "if-then" statements. For example, if a customer buys product A, then there is a high probability that they will also buy product B.

## General Exmaple

If a customer buys bread and milk, then there is a high probability that they will also buy butter.
This rule implies that customers who purchase bread and milk are likely to purchase butter as well. This insight can be valuable for the grocery store in terms of product placement, promotions, and recommendations to customers.

## Support

Support is a measure that indicates the frequency or popularity of a combination of items in a dataset. It quantifies how often a particular combination of items appears together in the dataset. Support is calculated as the ratio of the number of transactions containing the combination of items to the total number of transactions in the dataset.  

For example, if we have a dataset of 100 transactions, and the combination of items A and B appears together in 30 transactions, then the support for the combination (A, B) would be 30/100 = 0.3 or 30%.

## Confidence (from A to B)

Confidence measures the reliability or strength of an association rule. Specifically, it quantifies how often an item B is purchased when item A is purchased. Confidence is calculated as the ratio of the number of transactions containing both items A and B to the number of transactions containing item A.  

For example, if item A appears in 50 transactions, and out of those 50 transactions, item B also appears in 30 transactions, then the confidence from A to B would be 30/50 = 0.6 or 60%.

## Lift

Lift is a measure that assesses the strength of the association between two items (A and B) beyond what would be expected by chance. It compares the observed probability of items A and B appearing together to the expected probability of their co-occurrence if they were statistically independent. Lift is calculated as the ratio of the support for the combination of items (A, B) to the product of the individual supports of items A and B.
A lift value greater than 1 indicates a positive association or correlation between the items. The higher the lift value, the stronger the association.  

For example, if the support for the combination (A, B) is 0.3, and the individual supports for items A and B are 0.5 and 0.4 respectively, then the lift would be 0.3 / (0.5 * 0.4) = 1.5.

# Load data from BigQuery

In [3]:
def bq_connector():
    key_path = "C:/Users/HEN1/Projects/Instacart_Market_Basket_Analysis/keys/plucky-mile-327121-255163f80b63.json"
    credentials = service_account.Credentials.from_service_account_file(
        key_path, scopes=["https://www.googleapis.com/auth/cloud-platform"],
    )

    bqclient = bigquery.Client(credentials=credentials, project=credentials.project_id,)
    bqstorageclient = bigquery_storage.BigQueryReadClient(credentials=credentials)
    return bqclient, bqstorageclient

def bq_full_table_df(bqclient, bqstorageclient, table_name):
    sql_query = f"SELECT * FROM instacart.{table_name}"
    query_job = bqclient.query(sql_query)
    time.sleep(30)
    count =0 
    while query_job.state !='DONE':
        print("NOT DONE")
        if query_job.state =='PENDING':
            print(f"job from {table_name} is pending")
            break
        if query_job.state =='RUNNING':
            print(f"job from {table_name} is running")
            print(query_job.result())
            time.sleep(60)
            query_job.reload()
            time.sleep(10)
            count += 1
            if count>3:
                break
        else:
            print("may meet an error")
            break
    if query_job.state == 'DONE':
        print(f"successfully finished getting data from {table_name} table")
        df = query_job.to_dataframe(bqstorage_client=bqstorageclient, 
                                    progress_bar_type='tqdm_notebook',)
        print("successfully transferred to df")
        time.sleep(3)
    else:
        print("error")

    return df

In [4]:
bqclient, bqstorageclient = bq_connector()

In [4]:
aisles = bq_full_table_df(bqclient, bqstorageclient, 'aisles')
time.sleep(3)

successfully finished getting data from aisles table


Query is running:   0%|          |

Downloading:   0%|          |

successfully transferred to df


In [5]:
departments = bq_full_table_df(bqclient, bqstorageclient, 'departments')
time.sleep(60)

successfully finished getting data from departments table


Query is running:   0%|          |

Downloading:   0%|          |

successfully transferred to df


In [6]:
orders = bq_full_table_df(bqclient, bqstorageclient, 'orders')
time.sleep(60)

NOT DONE
job from orders is running
successfully finished getting data from orders table


Query is running:   0%|          |

Downloading:   0%|          |

successfully transferred to df


In [12]:
products = bq_full_table_df(bqclient, bqstorageclient, 'products')
time.sleep(60)

successfully finished getting data from products table


Query is running:   0%|          |

Downloading:   0%|          |

successfully transferred to df


In [5]:
order_products_prior = bq_full_table_df(bqclient, bqstorageclient, 'order_products_prior')
time.sleep(60)

NOT DONE
job from order_products_prior is running
successfully finished getting data from order_products_prior table


Query is running:   0%|          |

Downloading:   0%|          |

successfully transferred to df


In [9]:
order_products_train = bq_full_table_df(bqclient, bqstorageclient, 'order_products_train')

NOT DONE
job from order_products_train is running
successfully finished getting data from order_products_train table


Query is running:   0%|          |

Downloading:   0%|          |

successfully transferred to df


# Extract data from dataset

In [17]:
order_products_prior[['order_id', 'product_id']].head(20)

order_id  product_id
0    1091226       27735
1    1679699        1100
2    3172993       31404
3     490071        6187
4     400235        4809
5     583860       24184
6    3092797       37935
7     155898        2245
8     241827       33754
9    1495286       39275
10   1926282        3850
11   2041145       24004
12    536935       44560
13   1838605       37807
14    699367       47626
15   3076117       13457
16   1283874       49325
17   1627942       26209
18    470033        3957
19    895793       49683

In [6]:
transactions = order_products_prior.groupby('order_id')['product_id'].apply(list).tolist()


In [7]:
def generate_itemsets(transactions):
    itemsets = []
    num_transactions = len(transactions)

    # Generate all possible itemsets of length 2
    for transaction in tqdm(transactions, desc="Generating Itemsets"):
        pairs = list(combinations(transaction, 2))
        itemsets.extend(pairs)

    # Create a DataFrame from itemsets
    itemsets_df = pd.DataFrame(itemsets, columns=['itemA_id', 'itemB_id'])
    itemsets_df['count'] = 1

    return itemsets_df


def analyze_itemsets(itemsets_df, transactions):
    num_transactions = len(transactions)

    # Calculate supportAB for each itemset
    itemsets_df_groupby = itemsets_df.groupby(['itemA_id', 'itemB_id'], as_index=False)['count'].sum()
    itemsets_df_groupby['supportAB'] = itemsets_df_groupby['count'] / num_transactions
    print("Finished calculate supportAB")
    itemsets_df_groupby.drop(['count'], axis=1, inplace=True)

    # Sort itemsets based on supportAB in descending order
    itemsets_df_groupby.sort_values('supportAB', ascending=False, inplace=True)

    # Create a DataFrame to store the item counts for single items
    single_item_counts = pd.Series(transactions).explode().value_counts().reset_index()
    single_item_counts.columns = ['item_id', 'count']

    # Calculate supportA and supportB for the top itemsets
    output = pd.merge(itemsets_df_groupby, single_item_counts, left_on='itemA_id', right_on='item_id', how='left')
    output['supportA'] = output['count'] / num_transactions
    output.drop(['item_id', 'count'], axis=1, inplace=True)

    output = pd.merge(output, single_item_counts, left_on='itemB_id', right_on='item_id', how='left')
    output['supportB'] = output['count'] / num_transactions
    output.drop(['item_id', 'count'], axis=1, inplace=True)

    # Calculate confidenceAtoB, confidenceBtoA, and lift for the top itemsets
    output['confidenceAtoB'] = output['supportAB'] / output['supportA']
    output['confidenceBtoA'] = output['supportAB'] / output['supportB']
    output['lift'] = output['supportAB'] / (output['supportA'] * output['supportB'])

    return output


In [8]:
# Generate itemsets
itemsets_df = generate_itemsets(transactions)

Generating Itemsets: 100%|██████████| 3214874/3214874 [00:24<00:00, 129648.85it/s]


In [10]:

# Analyze top K itemsets and store the results in a DataFrame
top_itemsets_df = analyze_itemsets(itemsets_df, transactions)

Finished calculate supportAB


In [19]:
top_itemsets_df = top_itemsets_df.merge(products[['product_id', 'product_name']], left_on='itemA_id', right_on='product_id', how='left').rename(columns={'product_name': 'productA_name'})
top_itemsets_df = top_itemsets_df.merge(products[['product_id', 'product_name']], left_on='itemB_id', right_on='product_id', how='left').rename(columns={'product_name': 'productB_name'})


In [22]:
top_itemsets_df = top_itemsets_df.drop(columns=['product_id_x', 'product_id_y'])

In [43]:
pd.set_option('display.max_rows', 200)

In [23]:
top_itemsets_df.head(100)

itemA_id  itemB_id  supportAB  supportA  supportB  confidenceAtoB  \
0      13176     47209   0.009736  0.118030  0.066436        0.082488   
1      47209     13176   0.009655  0.066436  0.118030        0.145334   
2      13176     21137   0.009602  0.118030  0.082331        0.081352   
3      21137     13176   0.009568  0.082331  0.118030        0.116211   
4      24852     21137   0.008838  0.146993  0.082331        0.060123   
5      21137     24852   0.008630  0.082331  0.146993        0.104820   
6      24852     47766   0.008329  0.146993  0.054999        0.056663   
7      47766     24852   0.008280  0.054999  0.146993        0.150542   
8      24852     21903   0.008035  0.146993  0.075251        0.054663   
9      21903     24852   0.007951  0.075251  0.146993        0.105667   
10     13176     21903   0.007904  0.118030  0.075251        0.066968   
11     21903     13176   0.007764  0.075251  0.118030        0.103178   
12     24852     16797   0.006566  0.146993  0.044466        0.044669   
13     24852     47626   0.006485  0.146993  0.047485        0.044115   
14     21137     47209   0.006376  0.082331  0.066436        0.077447   
15     13176     27966   0.006369  0.118030  0.042632        0.053957   
16     47209     21137   0.006313  0.066436  0.082331        0.095021   
17     16797     24852   0.006259  0.044466  0.146993        0.140769   
18     47626     24852   0.006231  0.047485  0.146993        0.131229   
19     27966     13176   0.006230  0.042632  0.118030        0.146136   
20     21137     21903   0.005964  0.082331  0.075251        0.072434   
21     21903     21137   0.005898  0.075251  0.082331        0.078381   
22     21903     47209   0.005433  0.075251  0.066436        0.072193   
23     47209     21903   0.005424  0.066436  0.075251        0.081635   
24     24852     28204   0.005361  0.146993  0.027880        0.036471   
25     21137     27966   0.005299  0.082331  0.042632        0.064360   
26     27966     21137   0.005234  0.042632  0.082331        0.122781   
27     28204     24852   0.005197  0.027880  0.146993        0.186407   
28     24852     49683   0.005089  0.146993  0.030270        0.034620   
29     24852     26209   0.005066  0.146993  0.043743        0.034461   
30     24852     27845   0.004931  0.146993  0.042896        0.033545   
31     27845     24852   0.004911  0.042896  0.146993        0.114492   
32     24852     47209   0.004901  0.146993  0.066436        0.033339   
33     49683     24852   0.004895  0.030270  0.146993        0.161712   
34     26209     24852   0.004817  0.043743  0.146993        0.110128   
35     47209     24852   0.004811  0.066436  0.146993        0.072416   
36     21903     47766   0.004806  0.075251  0.054999        0.063868   
37     47766     21903   0.004802  0.054999  0.075251        0.087312   
38     24852     45066   0.004494  0.146993  0.024812        0.030576   
39     45066     24852   0.004342  0.024812  0.146993        0.174993   
40     47626     26209   0.004288  0.047485  0.043743        0.090307   
41     26209     47626   0.004236  0.043743  0.047485        0.096831   
42     27845     13176   0.004145  0.042896  0.118030        0.096632   
43     13176     27845   0.004092  0.118030  0.042896        0.034669   
44     47209     27966   0.004014  0.066436  0.042632        0.060417   
45     27966     47209   0.004009  0.042632  0.066436        0.094041   
46     24852      4920   0.003921  0.146993  0.025721        0.026674   
47     13176     22935   0.003890  0.118030  0.035282        0.032961   
48     47766     47626   0.003842  0.054999  0.047485        0.069847   
49     21137     47766   0.003767  0.082331  0.054999        0.045757   
50     47626     47766   0.003753  0.047485  0.054999        0.079046   
51     22935     13176   0.003753  0.035282  0.118030        0.106387   
52     21137     39275   0.003749  0.082331  0.031124        0.045537   
53     21137     27845   0.003748  0.082331  0.042896        0

In [44]:
top_itemsets_df[(top_itemsets_df['productA_name']=='Organic Strawberries') & (top_itemsets_df['supportAB'] > 0.0005)].sort_values(by='lift', ascending=False)

itemA_id  itemB_id  supportAB  supportA  supportB  confidenceAtoB  \
1854     21137     19706   0.000757  0.082331  0.004075        0.009192   
835      21137     38159   0.001182  0.082331  0.006626        0.014357   
1129     21137     38777   0.001004  0.082331  0.006764        0.012196   
2292     21137     26790   0.000665  0.082331  0.005017        0.008081   
1754     21137     26940   0.000781  0.082331  0.005999        0.009487   
1749     21137     17600   0.000784  0.082331  0.006100        0.009521   
285      21137     39928   0.001997  0.082331  0.015597        0.024259   
554      21137     43122   0.001482  0.082331  0.011729        0.018003   
25       21137     27966   0.005299  0.082331  0.042632        0.064360   
3153     21137     11782   0.000559  0.082331  0.004547        0.006789   
1350     21137       260   0.000911  0.082331  0.007434        0.011070   
2404     21137     45633   0.000649  0.082331  0.005331        0.007881   
2965     21137     25340   0.000578  0.082331  0.004759        0.007016   
52       21137     39275   0.003749  0.082331  0.031124        0.045537   
242      21137     22035   0.002128  0.082331  0.018562        0.025842   
1137     21137     22963   0.001001  0.082331  0.008829        0.012162   
2668     21137     13629   0.000612  0.082331  0.005419        0.007435   
2395     21137     26283   0.000650  0.082331  0.005881        0.007892   
3011     21137     34448   0.000573  0.082331  0.005241        0.006959   
437      21137     22825   0.001653  0.082331  0.015215        0.020073   
2595     21137     18370   0.000622  0.082331  0.005739        0.007552   
1683     21137     36695   0.000800  0.082331  0.007432        0.009717   
307      21137     10749   0.001909  0.082331  0.017881        0.023190   
2798     21137     10132   0.000597  0.082331  0.005611        0.007246   
1093     21137     42828   0.001023  0.082331  0.009702        0.012422   
735      21137      9839   0.001268  0.082331  0.012044        0.015403   
933      21137      7781   0.001119  0.082331  0.010641        0.013594   
2790     21137      3952   0.000598  0.082331  0.005705        0.007265   
135      21137     30391   0.002616  0.082331  0.025006        0.031770   
1705     21137      5212   0.000794  0.082331  0.007631        0.009646   
1873     21137     43768   0.000752  0.082331  0.007275        0.009132   
171      21137     19057   0.002412  0.082331  0.023380        0.029295   
295      21137     26604   0.001964  0.082331  0.019206        0.023851   
3144     21137     43713   0.000560  0.082331  0.005611        0.006797   
963      21137     38689   0.001105  0.082331  0.011093        0.013424   
831      21137     46906   0.001184  0.082331  0.012122        0.014379   
3142     21137     37158   0.000560  0.082331  0.005786        0.006801   
3728     21137     17461   0.000504  0.082331  0.005231        0.006124   
14       21137     47209   0.006376  0.082331  0.066436        0.077447   
145      21137      8277   0.002529  0.082331  0.026446        0.030720   
2788     21137     14678   0.000598  0.082331  0.006292        0.007269   
2569     21137     46969   0.000625  0.082331  0.006666        0.007594   
159      21137     40706   0.002453  0.082331  0.026208        0.029790   
324      21137     41950   0.001868  0.082331  0.019997        0.022688   
871      21137     18465   0.001156  0.082331  0.012456        0.014036   
2950     21137     29447   0.000580  0.082331  0.006283        0.007042   
1331     21137     30233   0.000916  0.082331  0.009986        0.011123   
261      21137     37646   0.002073  0.082331  0.022659        0.025185   
1810     21137     21405   0.000767  0.082331  0.008469        0.009321   
2781     21137      7948   0.000599  0.082331  0.006690        0.007280   
520      21137     43961   0.001529  0.082331  0.017223        0.018566   
1392     21137     19678   0.000895  0.082331  0.010120        0.010866   
904      21137     27344   0.00

In [35]:
top_itemsets_df[top_itemsets_df['lift'].between(0.99, 1.01, inclusive=False)].head()

C:\Users\HEN1\AppData\Local\Temp/ipykernel_7900/2209779414.py:1: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated infavour of `both` or `neither`.
  top_itemsets_df[top_itemsets_df['lift'].between(0.99, 1.01, inclusive=False)].head()


itemA_id  itemB_id  supportAB  supportA  supportB  confidenceAtoB  \
12      24852     16797   0.006566  0.146993  0.044466        0.044669   
65      21903     47626   0.003585  0.075251  0.047485        0.047640   
152     13176     44359   0.002488  0.118030  0.021180        0.021075   
208     24852     25890   0.002282  0.146993  0.015640        0.015524   
228     22035     13176   0.002192  0.018562  0.118030        0.118071   

     confidenceBtoA      lift                productA_name  \
12         0.147666  1.004576                       Banana   
65         0.075496  1.003263         Organic Baby Spinach   
152        0.117446  0.995054       Bag of Organic Bananas   
208        0.145900  0.992562                       Banana   
228        0.018569  1.000351  Organic Whole String Cheese   

                         productB_name  
12                        Strawberries  
65                         Large Lemon  
152         Organic Small Bunch Celery  
208  Boneless Skinless Chicken Breasts  
228             Bag of Organic Bananas

In [124]:
top_itemsets_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 55137021 entries, 0 to 55137020
Data columns (total 8 columns):
 #   Column          Dtype  
---  ------          -----  
 0   itemA_id        int64  
 1   itemB_id        int64  
 2   supportAB       float64
 3   supportA        float64
 4   supportB        float64
 5   confidenceAtoB  float64
 6   confidenceBtoA  float64
 7   lift            float64
dtypes: float64(6), int64(2)
memory usage: 3.7 GB


In [24]:
top_itemsets_df.supportAB.max()

0.009735995874177339

# Insights And Findings

1. The association rule between "Bag of Organic Bananas" and "Organic Hass Avocado" has a high support (0.009736) and confidence (0.082488), indicating that these two items are frequently purchased together. The lift value (1.241609) suggests that the occurrence of "Organic Hass Avocado" and the occurrence of "Bag of Organic Bananas" are strong correlated. This is a pretty straight forward finding, because people who are willing to buy organic banana are possible to buy organic avocado as carbohydrates for health reasons.

2. The association rule between "Organic Strawberries" and "Banana" has a relatively high support (0.008630) and confidence (0.104820). This suggests that these two items are often bought together by customers. The lift value (0.713092) indicates that the occurrence of "Banana" is slightly reduced when "Organic Strawberries" is purchased. From the data above, we can tell that people who buy organic strawberries are more willing to buy healthy food like natural or organic food, and not that willing to buy groceries like garlic. 

3. The association rule between "Banana" and "Strawberries" has a moderate support (0.006566) and confidence (0.044669). This indicates that there is a moderate association between these two items. The lift value (1.004576) close to 1 suggests that the occurrence of "Strawberries" has minimal impact on the likelihood of purchasing "Banana". The popular itemsets like "Organic Baby Spinach" and "Large Lemon", "Bags of Organic Bananas" and "Organic Small Bunch Celery", "Bag of Organic Bananas" and "Organic Small Bunch Celery", "Banana" and "Boneless Skinless Chicken Breasts", "Organic Whole String Cheese" and "Bag of Organic Bananas" are under the same situation. It implies that these items in itemsets are independent to each other. It is pretty hard to use common sense to explain this phenomenon, but we can suggest that the target customers for product A and product B are different.

4. The association rule between "Organic Hass Avocado" and "Organic Raspberries" has a relatively low support (0.004014) and confidence (0.060417). This suggests that these two items are not frequently purchased together. The lift value (1.417158) indicates that the occurrence of "Organic Raspberries" increases by 41.72% when "Organic Hass Avocado" is also purchased. We can use this finding to boost our sales by creating a "BUY X GET Y" discount for these two products. 

# Limitations

1. Due to the dataset is pretty large, this notebook only focus on itemset with 2 items. There are still many relationship worth to explore.

2. The Apriori algorithm focuses on identifying frequent itemsets but does not consider the order in which items appear within a transaction. It may overlook important sequential patterns or temporal dependencies present in the data.

3. Association rules generated by the Apriori algorithm provide information about co-occurrence patterns but lack insights into the strength or causality of relationships between items. It does not capture the complex interactions and dependencies that may exist between items.